File: c_files/raylib/single_file_programs/breakout.c
#include <assert.h>
#include <raylib.h>
#include <raymath.h>

static const int SCREEN_WIDTH  = 1280;
static const int SCREEN_HEIGHT = 720;

static const int   BRICK_ROWS   = 5;
static const int   BRICK_COLS   = 12;
static const float BRICK_HEIGHT = 30.0f;
static const float BRICK_GAP    = 5.0f;
static const float BRICK_MARGIN = 50.0f;

static void draw_centered_text(const char* text, const Rectangle bounds, const float font_size) {
    assert(text);

    const Font    font      = GetFontDefault();
    const float   spacing   = 2.0f;
    const Vector2 text_size = MeasureTextEx(font, text, font_size, spacing);
    const Vector2 pos       = (Vector2){
        bounds.x + (bounds.width - text_size.x) / 2.0f,
        bounds.y + (bounds.height - text_size.y) / 2.0f,
    };
    const Color tint = BLACK;

    DrawTextEx(font, text, pos, font_size, spacing, tint);
}

static inline float get_brick_width(void) {
    const float available_width
        = GetScreenWidth() - (BRICK_MARGIN * 2) - (BRICK_GAP * (BRICK_COLS - 1));

    return available_width / BRICK_COLS;
}

typedef struct {
    float x;
    float y;
} Brick;

typedef struct {
    Rectangle rect;
    Color     color;
    float     speed;
} Paddle;

static void paddle_reset(Paddle* paddle) {
    assert(paddle);

    paddle->rect.x = (GetScreenWidth() - paddle->rect.width) / 2.0f;
}

static void paddle_init(Paddle* paddle) {
    assert(paddle);

    const float width  = 100.0f;
    const float height = 15.0f;
    paddle->rect       = (Rectangle){0, GetScreenHeight() - (height * 2), width, height};
    paddle->color      = BLACK;
    paddle->speed      = 300.0f;

    paddle_reset(paddle);
}

static void paddle_update(Paddle* paddle, const float dt) {
    if (IsKeyDown(KEY_A) && paddle->rect.x > 0) {
        paddle->rect.x -= paddle->speed * dt;
    }
    if (IsKeyDown(KEY_D) && paddle->rect.x < GetScreenWidth() - paddle->rect.width) {
        paddle->rect.x += paddle->speed * dt;
    }
}

static void paddle_draw(const Paddle* paddle) {
    assert(paddle);

    DrawRectangleRec(paddle->rect, paddle->color);
}

typedef struct {
    const Paddle* paddle;
    Vector2       center;
    float         radius;
    Color         color;
    float         speed;
    float         speed_increment;
    Vector2       direction;
    bool          active;
} Ball;

static void ball_update_pos(Ball* ball) {
    assert(ball);

    ball->center = (Vector2){
        ball->paddle->rect.x + ball->paddle->rect.width / 2.0f,
        ball->paddle->rect.y - (ball->radius * 2),
    };
}

static void ball_reset(Ball* ball) {
    assert(ball);
    ball->speed     = 400.0f;
    ball->direction = (Vector2){GetRandomValue(0, 1) == 0 ? -1 : 1, -1};
    ball->active    = false;

    ball_update_pos(ball);
}

static void ball_init(Ball* ball, const Paddle* paddle) {
    assert(ball);
    assert(paddle);

    ball->paddle          = paddle;
    ball->radius          = 10.0f;
    ball->color           = RED;
    ball->speed_increment = 10.0f;

    ball_reset(ball);
}

static void ball_update(Ball* ball, const float dt) {
    assert(ball);

    // update pos
    if (!ball->active) {
        ball_update_pos(ball);
    }

    // activate ball
    if (IsKeyPressed(KEY_SPACE)) {
        ball->active = true;
    }

    if (ball->active) {
        ball->direction = Vector2Normalize(ball->direction);

        // move
        ball->center.x += ball->direction.x * ball->speed * dt;
        ball->center.y += ball->direction.y * ball->speed * dt;

        // set bounds
        if (ball->center.x < ball->radius) {
            ball->center.x = ball->radius;
            ball->direction.x *= -1;
        }

        if (ball->center.x > GetScreenWidth() - ball->radius) {
            ball->center.x = GetScreenWidth() - ball->radius;
            ball->direction.x *= -1;
        }

        if (ball->center.y < ball->radius) {
            ball->center.y = ball->radius;
            ball->direction.y *= -1;
        }
    }
}

static void ball_draw(const Ball* ball) {
    assert(ball);

    DrawCircleV(ball->center, ball->radius, ball->color);
}

int main(void) {
    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, "Breakout");

    int  lives     = 5;
    bool game_over = false;
    bool game_won  = false;

    const float brick_width = get_brick_width();
    const Color brick_color = GRAY;

    int   brick_size = BRICK_ROWS * BRICK_COLS;
    Brick bricks[brick_size];
    for (int i = 0; i < BRICK_ROWS; i++) {
        for (int j = 0; j < BRICK_COLS; j++) {
            const int x = j * (brick_width + BRICK_GAP) + BRICK_MARGIN;
            const int y = i * (BRICK_HEIGHT + BRICK_GAP) + BRICK_MARGIN;

            bricks[i * BRICK_COLS + j] = (Brick){x, y};
        }
    }

    Paddle paddle;
    paddle_init(&paddle);

    Ball ball;
    ball_init(&ball, &paddle);

    while (!WindowShouldClose()) {
        if (brick_size == 0) game_won = true;
        if (lives == 0) game_over = true;

        if (!game_over && !game_won) {
            const float dt = GetFrameTime();

            paddle_update(&paddle, dt);
            ball_update(&ball, dt);

            // collisions
            // ball vs paddle
            if (CheckCollisionCircleRec(ball.center, ball.radius, paddle.rect)) {
                ball.speed += ball.speed_increment * dt;
                ball.direction.y *= -1;
            }
            // ball vs bricks
            for (int i = 0; i < brick_size; i++) {
                const Brick* brick = &bricks[i];

                if (CheckCollisionCircleRec(
                        ball.center,
                        ball.radius,
                        (Rectangle){brick->x, brick->y, brick_width, BRICK_HEIGHT})) {
                    ball.direction.y *= -1;
                    bricks[i] = bricks[--brick_size];
                    break;
                }
            }

            // update lives
            if (ball.center.y > GetScreenHeight() + ball.radius) {
                lives--;
                paddle_reset(&paddle);
                ball_reset(&ball);
            }

        } else {
            if (IsKeyPressed(KEY_ENTER)) {
                lives     = 5;
                game_over = false;
                game_won  = false;

                brick_size = BRICK_ROWS * BRICK_COLS;
                for (int i = 0; i < BRICK_ROWS; i++) {
                    for (int j = 0; j < BRICK_COLS; j++) {
                        const int x = j * (brick_width + BRICK_GAP) + BRICK_MARGIN;
                        const int y = i * (BRICK_HEIGHT + BRICK_GAP) + BRICK_MARGIN;

                        bricks[i * BRICK_COLS + j] = (Brick){x, y};
                    }
                }
                paddle_init(&paddle);
                ball_init(&ball, &paddle);
            }
        }

        BeginDrawing();
        ClearBackground(RAYWHITE);

        // draw lives
        DrawText(TextFormat("%d", lives), 20, GetScreenHeight() - 60, 30, BLACK);

        if (!ball.active && !game_over && !game_won) {
            draw_centered_text("press SPACE to begin",
                               (Rectangle){0, 0, GetScreenWidth(), GetScreenHeight() + 200},
                               20.0f);
        }

        ball_draw(&ball);
        paddle_draw(&paddle);
        // draw bricks
        for (int i = 0; i < brick_size; i++) {
            const Brick* brick = &bricks[i];
            DrawRectangleRec((Rectangle){brick->x, brick->y, brick_width, BRICK_HEIGHT},
                             brick_color);
        }

        if (game_over || game_won) {
            const char* text = (game_over) ? "GAME OVER" : "YOU WIN";

            const float     width     = 300.0f;
            const float     height    = 150.0f;
            const Rectangle rect      = (Rectangle){(GetScreenWidth() - width) / 2.0f,
                                                    (GetScreenHeight() - height) / 2.0f,
                                                    width,
                                                    height};
            const Color     color     = BLACK;
            const float     thickness = 5.0f;

            draw_centered_text(text, rect, 30.0f);
            draw_centered_text("press ENTER to restart",
                               (Rectangle){0, 0, GetScreenWidth(), GetScreenHeight() + 200},
                               20.0f);

            // draw outline rect
            DrawRectangleLinesEx(rect, thickness, color);
        }

        EndDrawing();
    }

    CloseWindow();

    return 0;
}

File: python_files/raylib/single_file_programs/breakout.py
import random

import pyray as p

import utils

SCREEN_WIDTH: int = 1280
SCREEN_HEIGHT: int = 720

BRICK_ROWS: int = 5
BRICK_COLS: int = 10
BRICK_HEIGHT: float = 30.0
BRICK_GAP: float = 5.0
BRICK_MARGIN: float = 50.0


def get_brick_width() -> float:
    available_width = (
        p.get_screen_width() - (BRICK_MARGIN * 2) - (BRICK_GAP * (BRICK_COLS - 1))
    )

    return available_width / BRICK_COLS


class Brick:
    def __init__(self, x: float, y: float) -> None:
        self.rect: p.Rectangle = p.Rectangle(x, y, get_brick_width(), BRICK_HEIGHT)
        self.color: p.Color = p.GRAY

    def draw(self) -> None:
        p.draw_rectangle_rec(self.rect, self.color)


class BrickManager:
    def __init__(self) -> None:
        self.bricks: list[Brick] = []

        brick_width: float = get_brick_width()

        for i in range(BRICK_ROWS):
            for j in range(BRICK_COLS):
                x: float = j * (brick_width + BRICK_GAP) + BRICK_MARGIN
                y: float = i * (BRICK_HEIGHT + BRICK_GAP) + BRICK_MARGIN

                self.bricks.append(Brick(x, y))

    def draw(self) -> None:
        for brick in self.bricks:
            brick.draw()


class Paddle:
    def reset(self) -> None:
        self.rect.x = (p.get_screen_width() - self.rect.width) / 2.0

    def __init__(self) -> None:
        width: float = 100.0
        height: float = 15.0
        self.rect: p.Rectangle = p.Rectangle(
            0, p.get_render_height() - (height * 2), width, height
        )

        self.color: p.Color = p.BLACK
        self.speed: float = 300.0

        self.reset()

    def update(self, dt: float) -> None:
        if p.is_key_down(p.KeyboardKey.KEY_A) and self.rect.x > 0:
            self.rect.x -= self.speed * dt

        if (
            p.is_key_down(p.KeyboardKey.KEY_D)
            and self.rect.x < p.get_screen_width() - self.rect.width
        ):
            self.rect.x += self.speed * dt

    def draw(self) -> None:
        p.draw_rectangle_rec(self.rect, self.color)


class Ball:
    def update_pos(self) -> None:
        self.center = p.Vector2(
            self.paddle.rect.x + self.paddle.rect.width / 2.0,
            self.paddle.rect.y - (self.radius * 2),
        )

    def reset(self) -> None:
        self.update_pos()
        self.speed: float = 400.0
        self.direction: p.Vector2 = p.Vector2(random.choice([-1, 1]), -1)
        self.active: bool = False

    def __init__(self, paddle: Paddle) -> None:
        self.paddle: Paddle = paddle

        self.radius: float = 10.0
        self.color: p.Color = p.RED
        self.speed_increment: float = 10.0

        self.reset()

    def update(self, dt: float) -> None:
        if not self.active:
            self.update_pos()

        # activate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = True

        if self.active:
            self.direction = p.vector2_normalize(self.direction)

            # move
            self.center.x += self.direction.x * self.speed * dt
            self.center.y += self.direction.y * self.speed * dt

            # set bounds
            if self.center.x < self.radius:
                self.center.x = self.radius
                self.direction.x *= -1

            if self.center.x > p.get_screen_width() - self.radius:
                self.center.x = p.get_screen_width() - self.radius
                self.direction.x *= -1

            if self.center.y < self.radius:
                self.center.y = self.radius
                self.direction.y *= -1

    def draw(self) -> None:

        p.draw_circle_v(self.center, self.radius, self.color)


class GameManager:
    def reset(self) -> None:
        self.lives: int = 1
        self.game_over: bool = False
        self.game_won: bool = False

        self.brick_manager: BrickManager = BrickManager()
        self.paddle: Paddle = Paddle()
        self.ball: Ball = Ball(self.paddle)

    def __init__(self) -> None:
        self.reset()

    def update(self, dt: float) -> None:
        if len(self.brick_manager.bricks) == 0:
            self.game_won = True

        if self.lives == 0:
            self.game_over = True

        if not self.game_over and not self.game_won:
            self.paddle.update(dt)
            self.ball.update(dt)

            # collisions
            # self.ball vs self.paddle

            if p.check_collision_circle_rec(
                self.ball.center, self.ball.radius, self.paddle.rect
            ):
                self.ball.direction.y *= -1
                self.ball.speed += self.ball.speed_increment * dt

            # bricks vs self.ball
            for brick in self.brick_manager.bricks:
                if p.check_collision_circle_rec(
                    self.ball.center, self.ball.radius, brick.rect
                ):
                    self.ball.direction.y *= -1
                    self.brick_manager.bricks.remove(brick)
                    break

            # update self.lives
            if self.ball.center.y > p.get_screen_height() + self.ball.radius:
                self.lives -= 1
                self.paddle.reset()
                self.ball.reset()

        else:
            if p.is_key_pressed(p.KeyboardKey.KEY_ENTER):
                self.reset()

    def draw(self) -> None:
        # draw lives
        p.draw_text(str(self.lives), 20, p.get_screen_height() - 50, 30, p.BLACK)

        self.brick_manager.draw()

        if not self.ball.active and not self.game_over and not self.game_won:
            utils.draw_centered_text(
                utils.OriginType.TOP_LEFT,
                "press SPACE to begin",
                p.Rectangle(0, 0, p.get_screen_width(), p.get_screen_height() + 200),
                20,
            )

        self.ball.draw()
        self.paddle.draw()

        if self.game_over or self.game_won:
            width: float = 300.0
            height: float = 150.0
            rect: p.Rectangle = p.Rectangle(
                (p.get_screen_width() - width) / 2.0,
                (p.get_screen_height() - height) / 2.0,
                width,
                height,
            )

            text: str = "GAME OVER" if self.game_over else "YOU WIN"

            utils.draw_centered_text(utils.OriginType.TOP_LEFT, text, rect, 30)
            utils.draw_centered_text(
                utils.OriginType.TOP_LEFT,
                "press ENTER to restart",
                p.Rectangle(0, 0, p.get_screen_width(), p.get_screen_height() + 200),
            )

            p.draw_rectangle_lines_ex(rect, 5.0, p.BLACK)


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, "Breakout")

    game_manager: GameManager = GameManager()

    while not p.window_should_close():
        dt: float = p.get_frame_time()

        game_manager.update(dt)

        p.begin_drawing()
        p.clear_background(p.RAYWHITE)

        game_manager.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
